File: lxb/leetcode/BestTimeToBuyAndSellStock.cpp
http://oj.leetcode.com/problems/best-time-to-buy-and-sell-stock/
/*
Say you have an array for which the ith element is the price of a given stock on day i.

If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.
*/

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        
        int min = 0;
        int maxDiff = 0;
        int buy = 0;
        int sell = 0;
        
        for (int i=0; i< prices.size(); i++){
             if (prices[i] < prices[min]) min = i;
             if (prices[i] - prices[min] > maxDiff){
                 maxDiff = prices[i] - prices[min];
                 buy = min;
                 sell = i;
             }
        }
        
        return maxDiff;
    }

}

File: lxb/leetcode/java/BestTimeToBuyandSellStock.java
/*
Say you have an array for which the ith element is the price of a given stock on day i.
If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.
*/
//Focus: 1. Integer.MAX_VALUE, Integer.MIN_VALUE  2. boundary case (empty size, 1)  3. Java variable naming?
public class Solution {
    public int maxProfit(int[] prices) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        int min_so_far = Integer.MAX_VALUE;
        int profit = 0;
        for (int i=0; i<prices.length;++i){
            if (min_so_far > prices[i]){
                min_so_far = prices[i];
            }
            
            if (profit < prices[i] - min_so_far) profit = prices[i] - min_so_far;
        }
        
        return profit;
    }
}
